### DISCUSSION TITLE
subscribeRepos bidirectional error reporting

### DISCUSSION THREAD

**Discussion by DavidBuchanan314** [bluesky-social/atproto#1212] [Protocol (atproto)]
As-is, the server can return error messages and "#info" messages to the subscriber.

However, if the *subscriber* reaches an error condition, there is no easy way for the server to know about it - it fails silently.

It would be nice if the subscriber could also return error messages to the server, along the lines of:

"I did not process event seq=123 because the signature was invalid" (or whatever the reason may be)

This would be useful for diagnosing federation issues in-the-wild, especially when instances might be running unknown software versions (so you can't, for example, reproduce an issue in a test environment)

**Comment by bnewbold** [comment]
This is a pretty reasonable idea. I'd want to dig back in to WebSocket protocol details to see if there is an idiomatic way of doing this.

I'm not sure if they need to be "errors" (resulting in the connection closing) or if just any "warning" or "info" would work (resulting in a log-line or similar on the server side).

This is kind of a novel thing in that HTTP doesn't work this way (clients don't have a mechanism for reporting errors to servers). And on some level we want to keep the subscriptions uni-directional to keep the door open to other transports which might not have bi-directional streams, like Kafka. Things like error and info messages are sort of special cases for streams though, so could support those for just some transports.

A related idea is to have infrastructure services always have a DID, and optionally an atproto account. This would provide a clear persona in app.bsky itself that could be notified of problems, probably as a manual poke. Different from automated error reporting, but could help with debugging issues.

**Comment by snarfed** [comment]
In the meantime, while event streams are still unidirectional, do you all have any suggestions for debugging `subscribeRepos`? Eg if the AppView isn't seeing a post or other record, any tips on how we should debug whether it's due to rejected sig, failed lexicon validation, etc?

**Reply by dholms** [reply]
we'll want to add some debug routes to the appview to help with this (and other stuff)

maybe something like `com.atproto.debug.recordStatus` and it will let you know if it's current record at the provided uri, and if it's passed validation

just curious, have you run into an issue related to this?

**Reply by snarfed** [reply]
Yes! That sounds great. And yes, this is exactly where I am right now. I'm happily serving `subscribeRepos`, calling `requestCrawl`, the BGS is subscribing and receiving records, but `getProfile` on the AppView isn't finding my account. There are at least a few possible reasons, including bad sigs - I don't yet do low-S 🤷 - and other user error - maybe the profile path isn't `app.bsky.actor.profile/self`, or the record doesn't validate, or bad MST entry(ies), etc.

**Reply by snarfed** [reply]
As a stopgap, can we assume that if the BGS calls `subscribeRepos` with `cursor: n`, can we assume it's accepted blocks with sequence numbers up to `n-1`? up to `n`?

**Reply by snarfed** [reply]
I've switched to low-S signatures, created a new DID and profile record (and some posts), BGS has subscribed and received the new `app.bsky.actor.profile/self`, it now calls `subscribeRepos` with a cursor past that record, but AppView is still serving 400 `Profile not found` to both `getProfile?actor=did:plc:3iz6xduyjrwxqlkwdeqlea2n` and `getProfile?actor=arroba3.snarfed.org`.

Looks like this is also discussed in https://github.com/bluesky-social/atproto/issues/1291 and a bit in https://github.com/bluesky-social/pds/issues/6 by people using the official PDS.

**Reply by snarfed** [reply]
Here's the profile record, in case it helps: https://arroba-pds.appspot.com/xrpc/com.atproto.repo.getRecord?repo=did:plc:3iz6xduyjrwxqlkwdeqlea2n&collection=app.bsky.actor.profile&rkey=self

```json
{
  "cid": "bafyreiab4mla3v2lcvduqk3vub73gzpeknbklfdqurhmfgx6liln5gf5ki",
  "uri": "at://did:plc:3iz6xduyjrwxqlkwdeqlea2n/app.bsky.actor.profile/self",
  "value": {
    "$type": "app.bsky.actor.profile",
    "avatar": {
      "$type": "blob",
      "mimeType": "image/jpeg",
      "ref": {
        "$link": "https://snarfed.org/ryan_profile_square.jpg"
      },
      "size": 0
    },
    "banner": {
      "$type": "blob",
      "mimeType": "image/jpeg",
      "ref": {
        "$link": "https://snarfed.org/w/wp-content/uploads/2013/06/copy-cropped-iraq_bar.jpg"
      },
      "size": 0
    },
    "description": "https://arroba-pds.appspot.com/",
    "displayName": "Ryan Arroba"
  }
}
```

**Reply by snarfed** [reply]
> As a stopgap, can we assume that if the BGS calls `subscribeRepos` with `cursor: n`, can we assume it's accepted blocks with sequence numbers up to `n-1`? up to `n`?

I'm guessing the answer is no, we can't infer anything from the BGS's current cursor. If it expects blocks and commits to be immutable, as it should, it probably discarded any commits it didn't like and still uses the highest `seq` it's seen as its cursor.

**Reply by snarfed** [reply]
The idea in https://github.com/bluesky-social/atproto/discussions/1212#discussioncomment-6814673 to check the sandbox BGS firehose for my commits was a good one. Turns out _none_ of my commits are there. OK then. Now to figure out why.

**Reply by snarfed** [reply]
(Made some progress. One key stumbling block was that after one commit fails to ingest, any future commits with `prev` pointers that lead back to it, whether directly or indirectly, will also fail. I worked around that by dropping `prev` from my `subscribeRepos` payload. [Repo v3](https://atproto.com/blog/repo-sync-update), here I come! 😆)

**Reply by bnewbold** [reply]
A resource you might want to look at is the integration test setup in `indigo`: https://github.com/bluesky-social/indigo/tree/main/cmd/fakermaker

There is a docker-compose file which runs PLC+PDS+BGS locally, and then there are specific tests in golang which, eg, create content and subscribe to the BGS. The idea is that you could slot your own PDS implementation or whatever in to that pipeline and test it.

The coverage isn't particularly good right now, but this has been useful for debugging and rolling out changes to, eg, repo structure or event stream protocol.

**Comment by snarfed** [comment]
@DavidBuchanan314 [mentioned a good idea for this in Discord just now](https://discord.com/channels/1097580399187738645/1098725036917002302/1144274870914514954): subscribe to the BGS's firehose. If you create a record on your PDS and then see the BGS emit it in `subscribeRepos`, it got accepted.

**Comment by snarfed** [comment]
Another idea here could be a filtered log of the sandbox BGS's recent commit ingest failures, eg just the last hour or something. @ericvolp12 [shared a couple of these logs with me just now on Discord](https://discord.com/channels/1097580399187738645/1119608517725266153/1144321907152396388) and they were very helpful. Redact anything sensitive, of course, but I'm guessing there shouldn't be much of that.

**Reply by bnewbold** [reply]
A back-burner idea is to have a debug web interface where you can request a crawl; the service will connect and attempt to subscribe (and maybe backfill?) and then report any errors. Would pair with a reference PDS with fake test data that works (is valid), and a separate reference PDS with a bunch of corrupt/bad/invalid data.

**Comment by snarfed** [comment]
I ended up running a local PDS with a [local BGS](https://github.com/bluesky-social/indigo/tree/main/cmd/bigsky) to see its error logs, which is working pretty well so far.

```sh
go run ./cmd/bigsky --crawl-insecure-ws --plc-host https://plc.bsky-sandbox.dev

# run my PDS locally

# run this from my repo to call requestCrawl on the BGS
env BGS_HOST=localhost:2470 PDS_HOST=127.0.0.1:8080 python request_crawl.py
```

I also had to make a new DID that points to the local PDS in its DID doc, ie:

```json
  "services": {
    "atproto_pds": {
      "type": "AtprotoPersonalDataServer",
      "endpoint": "https://127.0.0.1:8080"
    }
  },
```

**Comment by snarfed** [comment]
Managed to get the local BGS to ingest an `app.bsky.actor.profile` record and emit it to its own `subscribeRepos`. Progress! Current status: getting this error:

```
ERROR	bgs	bgs/fedmgr.go:372	failed handling event from "127.0.0.1:8080" (11): handle user event failed: reading changed record from car slice: handling "": unrecognized type
```

from https://github.com/bluesky-social/indigo/blob/0f8fc7bb407b0c54a1be2b2f8e70cde4359d421a/lex/util/decoder.go#L72-L80 , on this `app.bsky.feed.post`:

```json
{
  "repo": "did:plc:uelmpyml3gy3s5tlcdpby4q7",
  "collection": "app.bsky.feed.post",
  "record": {
    "text": "hello world",
    "createdAt": "2023-08-23T19:01:00Z"
  }
}
```

**Reply by DavidBuchanan314** [reply]
You're missing the `$type` field! I believe the official PDS will add `$type` fields automatically when you use createRecord etc. (tbh, I think it shouldn't do that, but that's another issue heh), and this is something I've been meaning to add to picopds

**Reply by snarfed** [reply]
Hah, thank you! That definitely sounds likely, and fits the odd empty string in the error message. Good lead to try tomorrow.

**Reply by snarfed** [reply]
`$type` fixed it! Rookie mistake. Thanks again!

**Reply by bnewbold** [reply]
hmm, yeah, we do very little "magic" fixing of records in the reference PDS, but the `$type` does get added, and maybe we should remove that.
